running a single component without its dependencies

Hello!

In Quattor 13.6, I could run e.g. "ncm-ncd --configure accounts --nodeps --autodeps=0" and only ncm-accounts was executed.

In Quattor 14.2.1, "--autodeps" has been changed to a boolean: So I tried: ncm-ncd --configure accounts --nodeps --noautodeps

But I get this error (with error count 0) now:

> [INFO] NCM-NCD version 14.2.1 started by root at: Fri May 16 11:14:40 2014
> [INFO] Ignoring broken pre/post dependencies (--nodeps flag set)
> [ERROR] Unable to satisfy dependency spma
> [INFO] executing configure on components....
> # 
> 
> [OK]   0 errors, 0 warnings executing configure

The behaviour is the same if I run "ncm-ncd --configure accounts --noautodeps"

> [INFO] NCM-NCD version 14.2.1 started by root at: Fri May 16 11:18:06 2014
> [ERROR] Unable to satisfy dependency spma
> [INFO] executing configure on components....
> # 
> 
> [OK]   0 errors, 0 warnings executing configure

The option "--nodeps" makes no difference.

But how can I run a single component without triggering its dependencies?

The use case is the following:
During the KS post reboot phase, I would like to run ncm-accounts 
in order to create all accounts before ncm-spma installs certain RPMs. 
Some RPMs create users that does not exist with pseudo-random UIDs 
depending on the installation order. I would like to avoid these 
random UIDs and therefore need to create the accounts beforehand.
With Quattor 13.6 it was easy to achieve this goal. 

Making ncm-accounts a predependency of spma would be a solution for this case, but could be a problem in the future, if one would like to upgrade ncm-accounts (with incompatible schema changes).

Another use case is simply the wish to run a single compoment over and over again for testing and debugging. Especially the new yum-based spma is a lot slower than the old spma. Therefore I would be very useful to be able to run a single compoment ignoring all dependencies.

Best regards,
  Andreas

--- a/src/main/perl/ComponentProxyList.pm
+++ b/src/main/perl/ComponentProxyList.pm
@@ -458,13 +458,13 @@ sub missing_deps
         if (!exists($comps->{$pp})) {
             if (!$autodeps) {
                 $ec->ignore_error();
-                $self->error("Unable to satisfy dependency $pp");
-                return undef;
+                $self->warn("Not satifying dependency $pp");
+                return;
             }
             push(@deps, $pp);
         }
     }
-    return (1, @deps);
+    return (@deps);
 }
 
 sub get_proxies
@@ -480,13 +480,11 @@ sub get_proxies
             $self->info("Skipping component $comp");
             next;
         }
-        my ($ok, @deps) = $self->missing_deps($px, $comps);
+        my (@deps) = $self->missing_deps($px, $comps);
 
-        if ($ok) {
-            push(@pxs, $px);
-            push(@c, grep(!exists($comps->{$_}), @deps));
-            $comps->{$_} = 1 foreach @deps;
-        }
+	push(@pxs, $px);
+	push(@c, grep(!exists($comps->{$_}), @deps));
+	$comps->{$_} = 1 foreach @deps;
     }
     return @pxs;
 }
@@ -511,7 +509,6 @@ sub _getComponents
     }
 
     $self->skip_components(\%comps) if $self->{SKIP};
-
 
     my @comp_proxylist= $self->get_proxies(\%comps);
 

--- a/src/test/perl/get-components.t
+++ b/src/test/perl/get-components.t
@@ -40,7 +40,7 @@ $cl = NCD::ComponentProxyList->new($cfg, undef, qw(acomponent adep));
 is(scalar(@{$cl->{CLIST}}), 2, "A component and its dependency were loaded");
 
 $cl = NCD::ComponentProxyList->new($cfg, undef, qw(adep));
-is(scalar(@{$cl->{CLIST}}), 0,
+is(scalar(@{$cl->{CLIST}}), 1,
    "Dependencies are not loaded with autodeps == 0");
 
 $this_app->{CONFIG}->set('autodeps', 1);

